feat(fonts): add support for installing fonts from custom URLs

mecattaf opened the pull request:
### tldr;
- Add url-fonts source to download fonts from any public URL
- Support multiple archive formats (zip, tar.gz, tar.bz2, tgz)
- Support individual font files (otf, ttf)
- Add documentation and examples for the new feature

# Pull Request: Add URL Fonts Support to Fonts Module

## Description
This PR adds support for installing fonts from custom URLs to the fonts module, complementing the existing Nerd Fonts and Google Fonts support.

## What's New
- **New feature**: Install fonts from any public URL
- **Flexible formats**: Supports both archives (`.zip`, `.tar.gz`, `.tar.bz2`, `.tgz`) and individual font files (`.otf`, `.ttf`)
- **Custom naming**: Use `"Name|URL"` format to organize fonts by custom names
- **Automatic extraction**: Archives are automatically extracted and only font files are kept

## Use Cases
- Installing proprietary/licensed fonts from company repositories
- Using fonts from sources other than Nerd Fonts or Google Fonts
- Installing specific font versions from direct URLs
- Including custom or private fonts in images

## Example Configuration
```yaml
type: fonts
fonts:
  url-fonts:
    - "MyCustomFont|https://example.com/fonts/custom.otf"
    - "CorporateFonts|https://company.com/brand/fonts.zip"
    - "SpecificVersion|https://github.com/user/repo/releases/download/v1.0/font.tar.gz"
```

## Implementation Details
- Fonts are installed to `/usr/share/fonts/url-fonts/`
- Each font is organized in its own subdirectory based on the provided name
- Font cache is automatically updated after installation
- Proper error handling for failed downloads or invalid archives

## Documentation
- Updated README.md with comprehensive documentation
- Added examples for various use cases
- Updated module.yml with new examples
- Updated TypeSpec schema to include the new `url-fonts` property

mecattaf committed:
feat(fonts): add support for installing fonts from custom URLs

- Add url-fonts source to download fonts from any public URL
- Support multiple archive formats (zip, tar.gz, tar.bz2, tgz)
- Support individual font files (otf, ttf)
- Add documentation and examples for the new feature

gmpinder commented on modules/fonts/module.yml:
```diff
@@ -13,3 +13,6 @@ example: |
       - Open Sans
+    url-fonts:
+      - "CustomFont|https://example.com/my-font.otf"
+      - "CompanyFonts|https://company.com/fonts.zip"
```
Honest opinion, I'm personally not a fan of custom string patterns and would prefer something like

```yaml
type: fonts
url-fonts:
  - name: CustomFont
    url: https://example.com/my-font.otf
  - name: CompanyFonts
    url: https://company.com/fonts.zip
```

It's a little more verbose, but it makes it clearer what the values are used for. This way is also more useful in helping to validate the recipes. That's also what we've been trying to follow in other modules like the `dnf` module. 

I'd also like to hear the opinions of @xynydev and @fiftydinar on this.

mecattaf replied:
Sounds good

mecattaf replied:
@gmpinder implemented

gmpinder commented on modules/fonts/README.md:
Thanks for filling this out! Didn't realize we had missing docs here

mecattaf committed:
refactor: use object structure for url-fonts configuration

- Changed url-fonts from pipe-delimited strings to explicit objects
- Added UrlFont model with 'name' and 'url' properties
- Updated shell scripts to parse JSON objects
- Updated documentation and examples
- Follows BlueBuild module patterns for better clarity and validation

Addresses review feedback from @gmpinder

gmpinder commented on modules/fonts/module.yml:
```diff
@@ -1,15 +1,34 @@
+
+  /** URL to download the font from (supports .otf, .ttf files and archives). */
+  url: string;
+}
```
Looks like you accidentally copied the TSP file into this one

mecattaf replied:
Yes my bad thank you.

mecattaf committed:
Merge pull request #1 from blue-build/main

fix: package exclusion grammar and structure

mecattaf committed:
correct .yml file

mecattaf reviewed (commented):
done

mecattaf committed:
Merge branch 'blue-build:main' into add-url-fonts-support
